### PR TITLE
GH-123894: Simplify ascii decode using unaligned loads

### PR DESCRIPTION
**Pull request by rhpvorderman**
When working on https://github.com/python/cpython/pull/120212 I noticed that the code for ascii_decode could be worded much simpler. 

The current code prevents unaligned loads, but on x86-64 unaligned loads are not necessarily slower. By using memcpy the compiler is allowed to force aligned loads on platforms that require it, while allowing unaligned loads on platforms that support it. 
It simplifies the code and allows for easier manual unrolling if that is desired at a later point.

EDIT: I feel that this does not warrant a NEWS entry. So I haven't written one.

* Issue: gh-123894

### CODE REVIEW

File: Objects/unicodeobject.c [hunk 1]
```diff
@@ -4973,58 +4973,32 @@ static Py_ssize_t
...
-            _p += SIZEOF_SIZE_T;
-            q += SIZEOF_SIZE_T;
-        }
-        p = _p;
-        while (p < end) {
-            if ((unsigned char)*p & 0x80)
-                break;
-            *q++ = *p++;
+    Py_UCS1 *q = dest;
+    const char *size_t_end  = end - (SIZEOF_SIZE_T - 1);
```
Comment (picnixz) [inline Objects/unicodeobject.c:4977]: Not sure if we already discussed it but would `_Py_ALIGN_DOWN(end, SIZEOF_SIZE_T)` work?

File: Objects/unicodeobject.c [hunk 2]
```diff
@@ -4973,58 +4973,32 @@ static Py_ssize_t
...
+
+    while (p < size_t_end) {
+        /* Loading and storing a size_t using memcpy leads to unaligned loads
+           for platforms that can handle it. */
+        const size_t *restrict _p = (const size_t *)p;
+        size_t *restrict _q = (size_t *)q;
+        size_t value;
+        memcpy(&value, _p, SIZEOF_SIZE_T);
```
Comment (picnixz) [inline Objects/unicodeobject.c:4985]: How are performances affected by those multiple memcpy() calls?

**Review by picnixz — COMMENTED** [review]
While it's easier to read, I'm wondering whether the memcpy() calls could induce some overhead. We are copying many small parts and check for the `ASCII_CHAR_MASK` mask instead of one single large `memcpy()` and some pointer arithmetic behind.

File: Objects/unicodeobject.c (hunk 2 above)
Comment (rhpvorderman) [reply Objects/unicodeobject.c:4985]: Memcpy isn't called. Memcpy with a given size gets optimized to a single load.

Comment (rhpvorderman) [reply Objects/unicodeobject.c:4985]: Q.E.D: https://godbolt.org/z/a9xdjfoq4

File: Objects/unicodeobject.c (hunk 1 above)
Comment (rhpvorderman) [reply Objects/unicodeobject.c:4977]: Probably. Is using a macro here important?

File: Objects/unicodeobject.c (hunk 2 above)
Comment (picnixz) [reply Objects/unicodeobject.c:4985]: Ah I see the ASM yes. I'd suggest reformulating the comment by saying that "When supported, compilers optimize `memcpy()` into a single LOAD instruction if the number of bytes to copy is a literal value".

~~I confirmed it on GCC, but is it the case for MSVC & clang by the way? (NVM, I posted before seeing your reply)~~

**Comment by ZeroIntensity** [comment]
I wasn't aware of any precedent for using `size_t` in the codebase. I do see that it was there beforehand, but maybe that was a mistake -- we should be using `Py_ssize_t`, shouldn't we?

**Comment by rhpvorderman** [comment]
This can be closed as the CPython maintainers prefer aligned loads: https://github.com/python/cpython/pull/120212#discussion_r1751915168

**Pull request closed without merging**